TGIS: new module t.copy

metzm opened the pull request:
This new module `t.copy` creates a copy of a space-time dataset. It can also copy datasets from a different mapset to the current mapset. Optionally, maps registered in the dataset are also copied. By default, only a new space-time dataset is created.
This PR requires PR #1924 to work properly across mapsets.

veroandreo reviewed (changes requested):
Nice addition!

metzm commented on temporal/t.copy/t.copy.html:
```diff
@@ -0,0 +1,46 @@
+<h2>DESCRIPTION</h2>
+
+The purpose of <b>t.copy</b> is to create a copy of a space
+time raster dataset. The new space time dataset will be located in the
```
No, any stds. The word `raster` is a leftover from copying...

metzm commented:
@veroandreo Thanks for the instant review!

griembauer commented on temporal/t.copy/t.copy.html:
```diff
@@ -0,0 +1,46 @@
+
+<div class="code"><pre>
+g.mapsets mapset=modis_lst operation=add
+t.copy input=LST_Day_monthly type=strds output=LST_Day_monthly
```
Here, no `input=LST_Day_monthly@modis_lst` is required, because `modis_lst` is added to the mapset search path? (Just out of curiosity, I am not too familiar with `g.mapsets operation=add`)

griembauer commented on temporal/t.copy/testsuite/test_t_copy.py:
```diff
@@ -0,0 +1,163 @@
+        """Remove the temporary region"""
+        cls.del_temp_region()
+
+    def setUp(self):
```
isn't all of this already done in the setUpClass method?

metzm commented on temporal/t.copy/t.copy.py:
```diff
@@ -0,0 +1,241 @@
+                        % (maptype, row["name"], mapset)
+                    )
+
+                kwargs = {maptype: "%s,%s" % (row["id"], row["name"])}
```
The syntax of `g.copy` is `g.copy [raster=from,to] [raster_3d=from,to] [vector=from,to] ...`. Here I use kwargs to avoid an `if ... elif ...` construct. `maptype` is here one of raster, raster_3d, vector.

veroandreo commented:
I've downloaded the `modis_lst` mapset grass79 ready which contains the `LST_Day_monthly` strds, extracted it within NC location and from `user1` mapset tried to test `t.copy` as indicated in the html manual page, however it does not work as expected. This is what I do in mapset `user1`:

```
t.upgrade
g.mapsets mapset=modis_lst operation=add

GRASS nc_spm_08_grass7/user1:~ > t.copy input=LST_Day_monthly type=strds output=LST_Day_monthly
ERROR: Space time raster dataset <LST_Day_monthly> not found
GRASS nc_spm_08_grass7/user1:~ > t.list
----------------------------------------------
Space time raster datasets with absolute time available in mapset <modis_lst>:
LST_Day_monthly@modis_lst
  
GRASS nc_spm_08_grass7/user1:~ > t.copy input=LST_Day_monthly@modis_lst type=strds output=LST_Day_monthly
ERROR: Only maps from the same mapset can be registered
GRASS nc_spm_08_grass7/user1:~ > t.list
----------------------------------------------
Space time raster datasets with absolute time available in mapset <user1>:
LST_Day_monthly@user1
Space time raster datasets with absolute time available in mapset <modis_lst>:
LST_Day_monthly@modis_lst
```

It only gets copied with full name, not as indicated in the manual page. Could the error be removed? It's misleading.

In any case, the copied strds is empty:

```
GRASS nc_spm_08_grass7/user1:~ > t.rast.list LST_Day_monthly@user1
ERROR: Space time raster dataset <LST_Day_monthly@user1> is empty
```

Is this the desired behaviour? What's the point of making a copy if it will be empty? Should it read the maps from modis_lst? 

Then, when using `-c`, I get:
```
GRASS nc_spm_08_grass7/user1:~ > t.copy input=LST_Day_monthly@modis_lst type=strds output=LST_Day_monthly -c
Copying raster <MOD11B3.A2015001.h11v05.single_LST_Day_6km@modis_lst> to
current mapset as <MOD11B3.A2015001.h11v05.single_LST_Day_6km>
Traceback (most recent call last):
  File "/home/veroandreo/software/grass79-dev/dist.x86_64-pc-linux-gnu/scripts/t.copy", line 241, in <module>
    main()
  File "/home/veroandreo/software/grass79-dev/dist.x86_64-pc-linux-gnu/scripts/t.copy", line 162, in main
    if new_map.is_in_db(dbif, mapset):
TypeError: is_in_db() takes from 1 to 2 positional arguments but 3 were given
```

wenzeslaus commented:
Changing milestone to 8.2.0. Independent and thus can be backported if really needed in 8.0, but as a feature/addtion, it does not need to be in 8.0.0.

metzm commented:
@veroandreo I have updated the manual and fixed open issues with `t.copy`. I hope it works now as expected.

The issues mentioned by you:

> *  t.copy needs full STRDS name, not as in the example in the manual, otherwise it searches for the STRDS in the current mapset and yields error.
> *  t.copy with the full STRDS name copies the STRDS table but it is empty (t.info reports all None and t.rast.list nothing) and it yields the following error: ERROR: Only a map that was inserted in the temporal database can be registered in a space time dataset. This seems confusing to me as those maps are indeed in the tgis database, only in a diff mapset, which is the core of the problem here.
> *  t.copy only works as I'd had expected if -c flag is set, which does not avoid the data duplication that was the original motivation if I understood it correctly.

should be resolved by now.

I think that `t.copy` is an important but so far missing feature in TGIS because `g.copy` does not support copying space time datasets.

veroandreo commented:
Hi @metzm! Great news! I'll test this weekend :)

veroandreo commented:
All works as expected! Cool!

1. When using `t.copy` without full STRDS name, I get an error saying the strds is not found
```
GRASS nc_spm_08_grass7/user1:grass79-dev > t.copy input=LST_Day_monthly type=strds output=LST_Day_monthly
ERROR: Space time raster dataset <LST_Day_monthly@user1> not found
```
2. When I use the full strds name, only the database is copied, since when I list maps in the newly copied strds, mapset is the original one: 
```
GRASS nc_spm_08_grass7/user1:grass79-dev > t.copy input=LST_Day_monthly@modis_lst type=strds output=LST_Day_monthly
GRASS nc_spm_08_grass7/user1:grass79-dev > t.rast.list LST_Day_monthly@user1
name|mapset|start_time|end_time
MOD11B3.A2015001.h11v05.single_LST_Day_6km|modis_lst|2015-01-01 00:00:00|2015-02-01 00:00:00
MOD11B3.A2015032.h11v05.single_LST_Day_6km|modis_lst|2015-02-01 00:00:00|2015-03-01 00:00:00
MOD11B3.A2015060.h11v05.single_LST_Day_6km|modis_lst|2015-03-01 00:00:00|2015-04-01 00:00:00
MOD11B3.A2015091.h11v05.single_LST_Day_6km|modis_lst|2015-04-01 00:00:00|2015-05-01 00:00:00
MOD11B3.A2015121.h11v05.single_LST_Day_6km|modis_lst|2015-05-01 00:00:00|2015-06-01 00:00:00
``` 
3. When I use the -c flag to copy, also maps are copied to the current mapset:
```
t.copy -c --o input=LST_Day_monthly@modis_lst type=strds output=LST_Day_monthly
Copying raster maps to the current mapset...
Copying raster <MOD11B3.A2015001.h11v05.single_LST_Day_6km@modis_lst> to
current mapset as <MOD11B3.A2015001.h11v05.single_LST_Day_6km>
Copying raster <MOD11B3.A2015032.h11v05.single_LST_Day_6km@modis_lst> to
current mapset as <MOD11B3.A2015032.h11v05.single_LST_Day_6km>
...
GRASS nc_spm_08_grass7/user1:grass79-dev > t.rast.list LST_Day_monthly@user1
name|mapset|start_time|end_time
MOD11B3.A2015001.h11v05.single_LST_Day_6km|user1|2015-01-01 00:00:00|2015-02-01 00:00:00
MOD11B3.A2015032.h11v05.single_LST_Day_6km|user1|2015-02-01 00:00:00|2015-03-01 00:00:00
MOD11B3.A2015060.h11v05.single_LST_Day_6km|user1|2015-03-01 00:00:00|2015-04-01 00:00:00
MOD11B3.A2015091.h11v05.single_LST_Day_6km|user1|2015-04-01 00:00:00|2015-05-01 00:00:00
MOD11B3.A2015121.h11v05.single_LST_Day_6km|user1|2015-05-01 00:00:00|2015-06-01 00:00:00
...
```
